### DISCUSSION TITLE
[R] `tbl()` shouldn't use `DBI::dbExistsTable()`

### DISCUSSION THREAD

**Discussion by mgirlich** [duckdb/duckdb#8212] [Ideas]
Copied from [this dbplyr issue](https://github.com/tidyverse/dbplyr/issues/1333):

``` r
library(duckdb)
#> Warning: package 'duckdb' was built under R version 4.2.3
#> Loading required package: DBI
library(DBI)

conn <- dbConnect(duckdb(), tempfile())

dbExecute(conn, "create schema test")
#> [1] 0
dbWriteTable(conn, "test.cars", cars)
dbWriteTable(conn, DBI::Id(schema = "test", table = "cars"), cars)
dbGetQuery(conn, "select * from test.cars limit 1")
#>   speed dist
#> 1     4    2


dplyr::tbl(conn, dbplyr::in_schema("test", "cars"))
#> Error in (function (classes, fdef, mtable) : unable to find an inherited method for function 'dbExistsTable' for signature '"duckdb_connection", "dbplyr_schema"'
dplyr::tbl(conn, DBI::Id(schema = "test", table = "cars"))
#> # Source:   table<cars> [?? x 2]
#> # Database: DuckDB 0.8.2-dev77 [root@Darwin 21.6.0:R 4.2.2//var/folders/xx/01v98b6546ldnm1rg1_bvk000000gn/T//RtmpCC95Sf/file952722097a23]
#>    speed  dist
#>    <dbl> <dbl>
#>  1     4     2
#>  2     4    10
#>  3     7     4
#>  4     7    22
#>  5     8    16
#>  6     9    10
#>  7    10    18
#>  8    10    26
#>  9    10    34
#> 10    11    17
#> # ℹ more rows

dbDisconnect(conn, shutdown = T)
```

<sup>Created on 2023-07-10 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.2.2 (2022-10-31)
#>  os       macOS Big Sur ... 10.16
#>  system   x86_64, darwin17.0
#>  ui       X11
#>  language (EN)
#>  collate  en_US.UTF-8
#>  ctype    en_US.UTF-8
#>  tz       America/New_York
#>  date     2023-07-10
#>  pandoc   3.1.1 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version    date (UTC) lib source
#>  cli           3.6.1      2023-03-23 [1] CRAN (R 4.2.0)
#>  DBI         * 1.1.3      2022-06-18 [1] CRAN (R 4.2.0)
#>  dbplyr        2.3.2.9000 2023-07-02 [1] Github (tidyverse/dbplyr@6683f25)
#>  digest        0.6.32     2023-06-26 [1] CRAN (R 4.2.0)
#>  dplyr         1.1.2      2023-04-20 [1] CRAN (R 4.2.0)
#>  duckdb      * 0.8.1      2023-06-21 [1] https://duckdb.r-universe.dev (R 4.2.3)
#>  evaluate      0.21       2023-05-05 [1] CRAN (R 4.2.0)
#>  fansi         1.0.4      2023-01-22 [1] CRAN (R 4.2.0)
#>  fastmap       1.1.1      2023-02-24 [1] CRAN (R 4.2.0)
#>  fs            1.6.2      2023-04-25 [1] CRAN (R 4.2.0)
#>  generics      0.1.3      2022-07-05 [1] CRAN (R 4.2.0)
#>  glue          1.6.2      2022-02-24 [1] CRAN (R 4.2.0)
#>  htmltools     0.5.5      2023-03-23 [1] CRAN (R 4.2.0)
#>  knitr         1.43       2023-05-25 [1] CRAN (R 4.2.0)
#>  lifecycle     1.0.3      2022-10-07 [1] CRAN (R 4.2.0)
#>  magrittr      2.0.3      2022-03-30 [1] CRAN (R 4.2.0)
#>  pillar        1.9.0      2023-03-22 [1] CRAN (R 4.2.0)
#>  pkgconfig     2.0.3      2019-09-22 [1] CRAN (R 4.2.0)
#>  purrr         1.0.1      2023-01-10 [1] CRAN (R 4.2.0)
#>  R.cache       0.16.0     2022-07-21 [1] CRAN (R 4.2.0)
#>  R.methodsS3   1.8.2      2022-06-13 [1] CRAN (R 4.2.0)
#>  R.oo          1.25.0     2022-06-12 [1] CRAN (R 4.2.0)
#>  R.utils       2.12.2     2022-11-11 [1] CRAN (R 4.2.0)
#>  R6            2.5.1      2021-08-19 [1] CRAN (R 4.2.0)
#>  reprex        2.0.2      2022-08-17 [1] CRAN (R 4.2.0)
#>  rlang         1.1.1      2023-04-28 [1] CRAN (R 4.2.0)
#>  rmarkdown     2.22       2023-06-01 [1] CRAN (R 4.2.0)
#>  rstudioapi    0.14       2022-08-22 [1] CRAN (R 4.2.0)
#>  sessioninfo   1.2.2      2021-12-06 [1] CRAN (R 4.2.0)
#>  styler        1.10.1     2023-06-05 [1] CRAN (R 4.2.0)
#>  tibble        3.2.1      2023-03-20 [1] CRAN (R 4.2.0)
#>  tidyselect    1.2.0      2022-10-10 [1] CRAN (R 4.2.0)
#>  utf8          1.2.3      2023-01-31 [1] CRAN (R 4.2.0)
#>  vctrs         0.6.3      2023-06-14 [1] CRAN (R 4.2.0)
#>  withr         2.5.0      2022-03-03 [1] CRAN (R 4.2.0)
#>  xfun          0.39       2023-04-20 [1] CRAN (R 4.2.0)
#>  yaml          2.3.7      2023-01-23 [1] CRAN (R 4.2.0)
#> 
#>  [1] /Library/Frameworks/R.framework/Versions/4.2/Resources/library
#> 
#> ──────────────────────────────────────────────────────────────────────────────
```

</details>

I think that duckdb tries to be a bit too smart here. Imho instead of supporting reading from files via `dplyr::tbl(con, "data/userdata1.parquet")` this should be use an explicit SQL query, i.e. `dplyr::tbl(con, dplyr::sql("SELECT * FROM data/userdata1.parquet"))`.
This would simplify the `tbl.duckdb_connection()` method to:

```r
tbl.duckdb_connection <- function(src, from, cache = FALSE, ...) {
  if (cache) DBI::dbExecute(src, "PRAGMA enable_object_cache")
  NextMethod("tbl")
}
```

Using `ident_q()` for table identifiers is also basically deprecated in the dev version of dbplyr.

**Comment by rsund** [comment]
DuckDB has a concept of replacement scan and it is difficult to see why that should not be supported with dbplyr backend. Moreover, there are also other more common use cases where DuckDB functions are used as source tables with `tbl`.

In addition, user defined schemas have not been available or at least not in active use in DuckDB at the time the backend was implemented, so it is no surprise that there may be some issues with schemas.

Thanks for pointing out that `ident_q` is "deprecated" in the dbplyr dev version.

In any case, unfortunately, it seems that the suggested simplification can't be used, because then DuckDB functions will not work directly with `tbl` (as there will be extra quotes in the wrong place) meaning that a major feature of the tailored backend would be gone.

But certainly some modifications to the method could be done to keep the current functionality and to fix detected issues.

**Reply by mgirlich** [reply]
Let me add some background here. There currently are several classes that provide a table identifier:

* string
* `ident()` -> `<ident>`
* `in_schema()` -> `<dbplyr_ident>`
* `in_catalog()` -> `<dbplyr_catalog>`
* `DBI::Id()` -> `<Id>`
* `ident_q()` -> `<ident_q>`

and two ways to provide an SQL query to power a table:

* `sql()` -> `<sql>`
* `DBI::SQL()` -> `<SQL>`

This mess of classes is difficult to support and frequently caused bugs in dbplyr. To avoid these issues all these classes are now converted in `tbl_sql()` to an internal class `<dbplyr_table_ident>` (or `<sql>` for the latter two cases).
While working it was also clear that `ident_q()` is only meant as a fallback in case of bugs in dbplyr. This is why the next version of dbplyr will inform the user that `ident_q()` is only meant as fallback and should usually not be used.

Apart from actual tables I see two different types of tables:

1. A function, e.g. `duckdb_keywords()`
2. A file, e.g. `'data/userdata1.parquet'` (which must be wrapped in `'`, right?)

I think the interface is cleaner overall if `tbl(con, <string>)` always connects to a table and other functions are used to select from a function or a file (e.g. `db_function("duckdb_keywords()"`). What do you think? (I'm not quite sure yet how this should be passed to dbplyr then.)

**Reply by rsund** [reply]
Thanks for the background, I see that supporting all those classes is challenging.

For the background of DuckDB backend implementation for `tbl` it is probably useful to recognize that the actual problem with `tbl` in the DuckDB backend was that dbplyr always put quotes around the table name. As the use of replacement scans (which DuckDB recognizes automatically from e.g. file names) and functions such as `read_parquet`, `read_csv`, or `duckdb_keywords` instead of a traditional table were not working with quotes, there was a need for dropping "extra" quotes in those cases in order to get generated sql to work with DuckDB parser.

The easiest way to do that turned out to be to override (drop) quotes if the "table" was not a traditional table and the simplest way to check if that was the case was to use the available DBI-function to do that and remove quotes only if it was not a traditional table.

Later on, it turned out that in some cases users wanted to use `dbplyr::sql` with `tbl` directly, so "tables" that actually were of type <sql> also required a special handling.

So the purpose of using a tailored class in the backend was not to actually change any essential dbplyr functionality but simply to remove quotes from the table names in the generated sql if needed.

I guess that one alternative option for the implementation could be to always remove quotes from names of the tables with DuckDB. That should probably work as far as there is no need for using tables corresponding to reserved keywords. The challenge would still be to automatically detect if a table with name conflicting reserved keyword exists, but maybe manually inserted quotes in `tbl` or some function could to used to indicate a need for that.

Summa summarum, with DuckDB backend dbplyr's behaviour to "always" put quotes to the table names   is problematic and therefore must be selectively overridden to unleash the full power of DuckDB. I think that the real question here is: What would be the smartest way to do that if not the current implementation with "hacky" but simple class override?

**Reply by mgirlich** [reply]
In the end dbplyr calls `DBI::dbQuoteIdentifier()`. This suggests to me that

* if `tbl()` supports reading from files and functions directly, then `DBI::dbQuoteIdentifier(con, "data.parquet")`, `DBI::dbQuoteIdentifier(con, "read_parquet(['data.parquet'])")`, `DBI::dbQuoteIdentifier(con, "duckdb_keywords()")` recognizes them as files resp. as database function and do the corresponding quoting
* give the user an explicit way to read from a file or a function, e.g. `function_tbl()` and `file_tbl()`
